test/cpp/api/nn_utils.cpp

```diff
@@ -104,6 +106,64 @@ TEST_F(NNUtilsTest, ClipGradNorm) {
   }
 }
 
+// Check that clip_grad_norm_ raises an error if the norm of a gradient
+// is non-finite
+TEST_F(NNUtilsTest, ClipGradNormErrorIfNonfinite) {
+  std::vector<std::tuple<double, bool>> test_cases({
+    // scalar, is_grad_nonfinite
+    std::make_tuple(std::numeric_limits<double>::infinity(), true),
+    std::make_tuple(-std::numeric_limits<double>::infinity(), true),
+    std::make_tuple(std::numeric_limits<double>::quiet_NaN(), true),
+    std::make_tuple(2e22, false),
+    std::make_tuple(-2e22, false)
```

Comment:
As mentioned for the python side in my comment below, I think there should be a test case for finite values large enough to result in a non-finite total norm for the gradients.